### PR TITLE
109605 dynamo scalar ndarray pow gen

### LINKED ISSUE #109605: [dynamo] We are not generating code for x**2 when x is an ndarray
### 🐛 Describe the bug

In this case, we are just falling back to `aten.pow`. This is caused by us casting the 2 into an array.

Tiny repro
```python
import torch
import numpy as np

@torch.compile
def fn(x):
    return x ** 2

fn(np.arange(8))
```

I tagged this as a good first issue, but I don't have time to guide for how to solve it. Fixing this is a matter of debugging through dynamo and finding where are we casting the 2 into a tensor and fixing that. If you want to work on this, simply submit a fix and tag me as a reviewer.

As a side note, it looks like we just do this for int arrays, as for floating point arrays we do generate code.

### Versions

master

cc @mruberry @rgommers @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @chenyang78 @aakhundov @kadeng

### PR DESCRIPTION
**Pull request by Aashishthakur10**
Fixes #109605

Generated code before:
```
def call(args):
    arg0_1, = args
    args.clear()
    assert_size_stride(arg0_1, (8, ), (1, ))
    buf0 = empty_strided((), (), device='cpu', dtype=torch.int64)
    cpp_fused_lift_fresh_0(c_void_p(buf0.data_ptr()))
    # Source Nodes: [wrapped_pow], Original ATen: [aten.lift_fresh, aten.pow]
    buf1 = aten.pow(arg0_1, reinterpret_tensor(buf0, (8, ), (0, ), 0))
    del arg0_1
    del buf0
    buf2 = buf1
    assert_size_stride(buf2, (8, ), (1, ))
    del buf1
    return (buf2, )
```

Generated code now:
```
def call(args):
    arg0_1, = args
    args.clear()
    assert_size_stride(arg0_1, (8, ), (1, ))
    buf0 = empty_strided((8, ), (1, ), device='cpu', dtype=torch.int64)
    cpp_fused_pow_0(c_void_p(arg0_1.data_ptr()), c_void_p(buf0.data_ptr()))
    del arg0_1
    return (buf0, )
```
@lezcano What would be a good way to add a test for this? 

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov @ColinPeppler

### CODE REVIEW

File: torch/_numpy/_ufuncs.py [hunk 1]
```diff
@@ -97,7 +97,11 @@ def cast(x, dtype):
         elif isinstance(x1, torch.Tensor) and isinstance(x2, torch.Tensor):
             dtype = _dtypes_impl.result_type_impl(x1, x2)
             x1, x2 = _util.typecast_tensors((x1, x2), dtype, casting)
-        else:
+        elif not (
+            any(isinstance(x, torch.Tensor) for x in [x1, x2])
+            and any(_dtypes_impl.is_scalar(x) for x in [x1, x2])
+        ):
+            # If one is already scalar and the other is tensor, no need for casting the scalar.
```
Comment (lezcano) [inline torch/_numpy/_ufuncs.py:104]: This is not right. NEP 50 is exactly to handle the tensor <> scalar case. A proper fix would go through checking in what cases we can avoid casting the scalar into a tensor.

No need to add a test, as this is thoroughly testesd in the codebase. In particular, the test you want to make sure that passes locally is `test/torch_np/test_nep50_examples.py` (and more generally, the whole `test/torch_np`)

Comment (Aashishthakur10) [reply torch/_numpy/_ufuncs.py:104]: I see, thanks for the help, I will update the code!

Comment (Aashishthakur10) [reply torch/_numpy/_ufuncs.py:104]: Hey @lezcano, the way I am trying to handle it now is inside is by doing a comparison between the weak and the not weak data types like this `weak_dtype!=dt`. The issue is that some of the NEP 50 functions like `arctan2` require tensors for both arguments. So, I am confused how to handle that without explicitly checking for functions, which doesn't seem right to do either. Any suggestions on how to proceed? Should I be fine tuning my checks a little more?

Comment (Aashishthakur10) [reply torch/_numpy/_ufuncs.py:104]: Hi @lezcano , 
I'm currently comparing the weak and not weak data types using weak_dtype != dt. However, an issue is that certain NEP 50 functions, such as `arctan2`, which require tensors for both arguments. Thus I am uncertain about how to handle this situation without resorting to explicit function checks, which doesn't seem like the right approach either. Do you have any suggestions on how to proceed? Should I consider refining my checks further?

Comment (Aashishthakur10) [reply torch/_numpy/_ufuncs.py:104]: That makes a lot of sense, and along the lines of what I was thinking, thank you for confirming!

**Review by lezcano — COMMENTED** [review]
Does this change the code generated by pow?

At any rate, can you add a regression test in testorchinductor.py?

**Comment by Aashishthakur10** [comment]
Yeah, the generated code is as follows: 
```
def call(args):
    arg0_1, = args
    args.clear()
    assert_size_stride(arg0_1, (8, ), (1, ))
    buf0 = empty_strided((8, ), (1, ), device='cpu', dtype=torch.int64)
    cpp_fused_pow_0(c_void_p(arg0_1.data_ptr()), c_void_p(buf0.data_ptr()))
    del arg0_1
    return (buf0, )

```
Sure, I will add a regression test there. Do you have any examples of the same? I will look into that file as well for it!

**Comment by lezcano** [comment]
Ah, I misread your fix. Yep, this is 100% correct.

For the regression test, have a look at `test/inductor/test_torchinductor.py` and in particular, the tests that inspect the generated code.

**Comment by lezcano** [comment]
nit. Usually it's best to base your PRs off `viable/strict`. This branch has the invariant that its tip always passes all CI (modulo flakiness and so on), so you'll minimise the number of unrelated reds you see in CI

File: test/inductor/test_torchinductor.py [hunk 2]
```diff
@@ -7095,6 +7095,27 @@ def fn(tensor, index, source):
...
+
+        x = torch.randn(8, device=self.device)
+        pow_opt = torch._dynamo.optimize("inductor")(pow)
+
+        actual, source_code = run_and_get_code(pow_opt, x)
+        expect = aten.pow(x, 2)
+
+        # assert that the functionality matches aten and self prior to being compiled by dynamo
+        self.assertEqual(expect, actual)
+        self.assertEqual(actual, pow(x))
```
Comment (lezcano) [inline test/inductor/test_torchinductor.py:7110]:
```suggestion
        self.assertEqual(actual, pow(x))
```
No need to execut it twice.

File: test/inductor/test_torchinductor.py [hunk 3]
```diff
@@ -7095,6 +7095,27 @@ def fn(tensor, index, source):
...
+
+        pattern = "aten.pow"
+        for code in source_code:
+            self.assertIsNone(
+                re.search(pattern, code),
+                msg="Encountered an unexpected fallback to 'aten pow' in dynamo compiled code",
+            )
```
Comment (lezcano) [inline test/inductor/test_torchinductor.py:7117]: can be golfed as
```suggestion
            self.assertIsNone(
                all("aten.pow" not in code for code in source_code"),
                msg="Encountered an unexpected fallback to 'aten pow' in dynamo compiled code",
            )
```
We also avoid using regex when possible, as they are incredibly slow

File: test/inductor/test_torchinductor.py [hunk 4]
```diff
@@ -7095,6 +7095,27 @@ def fn(tensor, index, source):
             ),
         )
 
+    def test_torch_dynamo_codegen_pow(self):
+        def pow(x):
+            return x**2
+
+        x = torch.randn(8, device=self.device)
```
Comment (lezcano) [inline test/inductor/test_torchinductor.py:7102]: We want to test NumPy here not PyTorch!

File: test/inductor/test_torchinductor.py [hunk 5]
```diff
@@ -7095,6 +7095,27 @@ def fn(tensor, index, source):
...
+        pow_opt = torch._dynamo.optimize("inductor")(pow)
```
Comment (lezcano) [inline test/inductor/test_torchinductor.py:7103]: nit.
```suggestion
        pow_opt = torch.compile(pow)
```

**Review by lezcano — COMMENTED** [review]
We are testing NumPy not PyTorch! Otherwise this should be good to go :)

File: test/inductor/test_torchinductor.py (hunk 4 above)
Comment (lezcano) [reply test/inductor/test_torchinductor.py:7102]: Put it in  `test_misc.py` better, as we don't want to test it for CUDA. There are a few other numpy tests there as well

Comment (Aashishthakur10) [reply test/inductor/test_torchinductor.py:7102]: My bad, brain fart moment right there!

**Review by lezcano — APPROVED** [review]
Great, thank you!

**Comment by huydhn** [comment]
@pytorchbot drci

(Please ignore this, I'm testing Dr. CI)

**Pull request closed without merging**